cmd/zoekt-sourcegraph-indexserver/main.go

```diff
@@ -1049,6 +1076,177 @@ func (s *Server) forceIndex(id uint32) (string, error) {
 	return fmt.Sprintf("Indexed %s with state %s", args.String(), state), nil
 }
 
+// Index implements the gRPC method for indexing a repository. If the repository
+// exists in the trash, it will attempt to recover it before indexing. The
+// method returns metadata about the indexed repository including branches and
+// and the time of indexing.  The metadata might be empty if we fail to read the
+// metadata after the index or if the repository is empty.
+//
+// Possible GRPC error codes:
+// - InvalidArgument: returned when the index options are missing or invalid
+// - AlreadyExists: returned when the repository is already being indexed by another process
+// - Internal: returned when indexing fails
+// - Canceled: returned when the context is canceled while waiting for an index slot
+func (s *Server) Index(ctx context.Context, req *indexserverv1.IndexRequest) (*indexserverv1.IndexResponse, error) {
+	return s.indexGRPC(ctx, req, s.index)
+}
+
+// indexGRPC takes an index func which is used for testing. For production GRPC calls, use
+// Server.Index instead
+func (s *Server) indexGRPC(ctx context.Context, req *indexserverv1.IndexRequest, indexFunc func(*indexArgs) (indexState, error)) (*indexserverv1.IndexResponse, error) {
+	// Validate request
+	if req.Options == nil {
+		return nil, status.Error(codes.InvalidArgument, "index options are required")
+	}
+
+	// Try to acquire semaphore, but respect context cancellation
+	select {
```

Comment:
A couple thoughts:
* Maybe we should set a default overall cancellation to protect against cases where clients don't set proper cancellation? This feels like a reasonable "defense in depth" practice. We could use the overall default indexing timeout. 
* It would be good to have a metric for how long requests wait on the semaphore.